File: backjoon/javascript/단계별 풀이/15. 동적계획법/계단 오르기.js
const input = require("fs").readFileSync(`/dev/stdin`, `utf8`).trim().split(`\n`);

let len = + input.shift();
let stair = input.map((num) => +num);
let max = [];

function maxNum(a, b) {
    return (a > b ? a : b);
}


// 3가지 경우 ( 한/두계단 씩, 연속 3개 밟으면 안되고, 마지막 계단은 꼭 밟아야해 )
max[0] = stair[0]; // 첫계단
max[1] = maxNum(stair[0] + stair[1], stair[1]); // 1+2계단 vs 바로 2번째 계단
max[2] = maxNum(stair[0] + stair[2], stair[1] + stair[2]); // 1+3계단 vs 2+3계단

for (let i = 3; i < len; i++) {
    max[i] = maxNum(stair[i] + stair[i - 1] + max[i - 3], stair[i] + max[i - 2]); // 마지막 계단 경우의 수를 계산
}

console.log(max[len - 1]);

File: backjoon/python/dynamic programming/계단 오르기.py
import sys
input = sys.stdin.readline

N = int(input())
dp = [0 for _ in range(N+3)]
arr = [0 for _ in range(N+3)]
for k in range(1,N+1):
    arr[k] = int(input())


dp [1] = arr[1]
dp [2] = arr[1] + arr[2]
dp [3] = max(arr[1] + arr[3] ,arr[2] + arr[3])


for i in range(4, N+1):
    dp[i] = max( dp[i-3] + arr[i-1] + arr[i] ,  dp[i-2] + arr[i] ) 
print(dp[N])


import sys
input = sys.stdin.readline
n=int(input())
stairs=[0]*300  # 변수 n 입력시 런타임 에러떠

for i in range(n):
  stairs[i]=int(input())

d=[0]*300
d[0]=stairs[0]
d[1]=stairs[0]+stairs[1]
d[2]=max(stairs[2]+stairs[0], stairs[2]+stairs[1])

for i in range(3,n):
  d[i]=max(stairs[i]+d[i-2], stairs[i]+stairs[i-1]+d[i-3])
print(d[n-1])


from sys import stdin

a,b,c = 0,0,0

n = int(stdin.readline())
for _ in range(n):
    stair = int(stdin.readline())
    d_0,d_1,d_2 = max(b,c), a+stair, b+stair
    a,b,c = d_0,d_1,d_2

print(max(d_2,d_1))
